core/src/main/scala/org/scalafmt/util/TokenOps.scala

```diff
@@ -52,10 +53,14 @@ object TokenOps {
     longHash
   }
 
-  def shouldGet2xNewlines(tok: FormatToken): Boolean = {
+  def shouldGet2xNewlines(tok: FormatToken, style: ScalafmtConfig): Boolean = {
     !isDocstring(tok.left) && {
       val newlines = newlinesBetween(tok.between)
       newlines > 1 || (isDocstring(tok.right) && !tok.left.is[Comment])
+    } || {
+      style.newlines.beforeTopLevelStatements &&
+        tok.between.count(_.is[KwNew]) < 2 &&
+        (tok.right.is[KwObject] || tok.right.is[KwDef] || tok.right.is[KwDef] || tok.right.is[KwCase] || tok.right.is[KwPackage])
```

Comment:
FYI, this won't cover modifiers like `final class` or `sealed trait`. Depending on your use-case, you might want to support that. There's a `owners: Token => Tree` available in Router that you can use to get the tree parent of the token. Note that the owner of modifiers like `lazy` is the Mod tree while the owner of `class` for exampl is Defn.Class.

btw, KwDef appears twice.